File: parser.cpp
// system includes
#include <map>
#include <memory>

// includes
#include "parser.h"
#include "generic_protocol/parser.h"
#include "basic_objects/parser.h"
#include "basic_parser/parser.h"
#include "basic_parser/malformed_request.h"
#include "validator.h"
#include "request_type_parser.h"

namespace user_reg_protocol
{

namespace parser
{

using ::basic_parser::parser::get_value_or_throw;
using ::basic_parser::parser::get_value_or_throw_t;

// enums

// objects

void get_value_or_throw( User * res, const std::string & prefix, const generic_request::Request & r )
{
    ::basic_objects::parser::get_value_or_throw( & res->gender, prefix + ".GENDER", r );
    get_value_or_throw( & res->login, prefix + ".LOGIN", r );
    get_value_or_throw( & res->last_name, prefix + ".LAST_NAME", r );
    get_value_or_throw( & res->first_name, prefix + ".FIRST_NAME", r );
    ::basic_objects::parser::get_value_or_throw( & res->email, prefix + ".EMAIL", r );
    get_value_or_throw( & res->phone, prefix + ".PHONE", r );
    ::basic_objects::parser::get_value_or_throw( & res->birthday, prefix + ".BIRTHDAY", r );
}

// base messages

void get_value_or_throw( Request * res, const generic_request::Request & r )
{
    // no base class
}

void get_value_or_throw( BackwardMessage * res, const generic_request::Request & r )
{
    // base class
    ::generic_protocol::parser::get_value_or_throw( static_cast<generic_protocol::BackwardMessage*>( res ), r );

}

// messages

void get_value_or_throw( RegisterUserRequest * res, const generic_request::Request & r )
{
    // base class
    parser::get_value_or_throw( static_cast<Request*>( res ), r );

    get_value_or_throw( & res->lead, "LEAD", r );
}

void get_value_or_throw( RegisterUserResponse * res, const generic_request::Request & r )
{
    // base class
    parser::get_value_or_throw( static_cast<BackwardMessage*>( res ), r );

}

void get_value_or_throw( ConfirmRegistrationRequest * res, const generic_request::Request & r )
{
    // base class
    parser::get_value_or_throw( static_cast<Request*>( res ), r );

    get_value_or_throw( & res->registration_key, "REGISTRATION_KEY", r );
}

void get_value_or_throw( ConfirmRegistrationResponse * res, const generic_request::Request & r )
{
    // base class
    parser::get_value_or_throw( static_cast<BackwardMessage*>( res ), r );

}

// to object

Object * to_RegisterUserRequest( const generic_request::Request & r )
{
    std::unique_ptr<RegisterUserRequest> res( new RegisterUserRequest );

    get_value_or_throw( res.get(), r );

    validator::validate( * res );

    return res.release();
}

Object * to_RegisterUserResponse( const generic_request::Request & r )
{
    std::unique_ptr<RegisterUserResponse> res( new RegisterUserResponse );

    get_value_or_throw( res.get(), r );

    validator::validate( * res );

    return res.release();
}

Object * to_ConfirmRegistrationRequest( const generic_request::Request & r )
{
    std::unique_ptr<ConfirmRegistrationRequest> res( new ConfirmRegistrationRequest );

    get_value_or_throw( res.get(), r );

    validator::validate( * res );

    return res.release();
}

Object * to_ConfirmRegistrationResponse( const generic_request::Request & r )
{
    std::unique_ptr<ConfirmRegistrationResponse> res( new ConfirmRegistrationResponse );

    get_value_or_throw( res.get(), r );

    validator::validate( * res );

    return res.release();
}

// to forward message

basic_parser::Object* to_forward_message( const generic_request::Request & r )
{
    auto type = detect_request_type( r );

    typedef request_type_e KeyType;

    typedef Object* (*PPMF)( const generic_request::Request & r );

#define HANDLER_MAP_ENTRY(_v)       { KeyType::_v,    & to_##_v }

    static const std::map<KeyType, PPMF> funcs =
    {
        HANDLER_MAP_ENTRY( RegisterUserRequest ),
        HANDLER_MAP_ENTRY( RegisterUserResponse ),
        HANDLER_MAP_ENTRY( ConfirmRegistrationRequest ),
        HANDLER_MAP_ENTRY( ConfirmRegistrationResponse ),
    };

#undef HANDLER_MAP_ENTRY

    auto it = funcs.find( type );

    if( it != funcs.end() )
        return it->second( r );

    return generic_protocol::parser::to_forward_message( r );
}

using basic_parser::MalformedRequest;

request_type_e  detect_request_type( const generic_request::Request & r )
{
    std::string cmd;

    if( r.get_value( "CMD", cmd ) == false )
        throw MalformedRequest( "CMD is not defined" );

    return RequestTypeParser::to_request_type( cmd );
}

} // namespace parser

} // namespace user_reg_protocol

File: parser.php
<?php

namespace user_reg_protocol;


// base include
require_once __DIR__.'/../generic_protocol/parser.php';
// includes
require_once __DIR__.'/../basic_objects/parser.php';
require_once __DIR__.'/../basic_parser/parser.php';

// enums

// objects

function parse__User( & $csv_arr, & $offset )
{
    $res = new User;

    $res->gender = \basic_objects\parse__gender_e( $csv_arr, $offset );
    $res->login = \basic_parser\parse__string( $csv_arr, $offset );
    $res->last_name = \basic_parser\parse__string( $csv_arr, $offset );
    $res->first_name = \basic_parser\parse__string( $csv_arr, $offset );
    $res->email = \basic_objects\parse__Email( $csv_arr, $offset );
    $res->phone = \basic_parser\parse__string( $csv_arr, $offset );
    $res->birthday = \basic_objects\parse__Date( $csv_arr, $offset );

    return $res;
}

// base messages

function parse__Request( & $res, & $csv_arr, & $offset )
{
    // no base class

}

function parse__BackwardMessage( & $res, & $csv_arr, & $offset )
{
    // base class
    \generic_protocol\parse__BackwardMessage( $res, $csv_arr, $offset );


}

// messages

function parse__RegisterUserRequest( & $csv_arr )
{
    $res = new RegisterUserRequest;

    $offset = 1;

    // base class
    parse__Request( $res, $csv_arr, $offset );

    $res->lead = parse__User( $csv_arr, $offset );
    $res->password = \basic_parser\parse__string( $csv_arr, $offset );

    return $res;
}

function parse__RegisterUserResponse( & $csv_arr )
{
    $res = new RegisterUserResponse;

    $offset = 1;

    // base class
    parse__BackwardMessage( $res, $csv_arr, $offset );


    return $res;
}

function parse__ConfirmRegistrationRequest( & $csv_arr )
{
    $res = new ConfirmRegistrationRequest;

    $offset = 1;

    // base class
    parse__Request( $res, $csv_arr, $offset );

    $res->registration_key = \basic_parser\parse__string( $csv_arr, $offset );

    return $res;
}

function parse__ConfirmRegistrationResponse( & $csv_arr )
{
    $res = new ConfirmRegistrationResponse;

    $offset = 1;

    // base class
    parse__BackwardMessage( $res, $csv_arr, $offset );


    return $res;
}

// generic

class Parser extends \generic_protocol\Parser
{

protected static function parse_csv_array( $csv_arr )
{
    if( sizeof( $csv_arr ) < 1 )
        return NULL;

    $handler_map = array(
        // messages
        'user_reg_protocol/RegisterUserRequest'         => 'parse__RegisterUserRequest',
        'user_reg_protocol/RegisterUserResponse'         => 'parse__RegisterUserResponse',
        'user_reg_protocol/ConfirmRegistrationRequest'         => 'parse__ConfirmRegistrationRequest',
        'user_reg_protocol/ConfirmRegistrationResponse'         => 'parse__ConfirmRegistrationResponse',
    );

    $type = $csv_arr[0][0];

    if( array_key_exists( $type, $handler_map ) )
    {
        $func = '\\user_reg_protocol\\' . $handler_map[ $type ];
        return $func( $csv_arr[0] );
    }

    return \generic_protocol\Parser::parse_csv_array( $csv_arr );
}

}

// namespace_end user_reg_protocol


?>
